src/main/java/camp/woowak/lab/infra/cache/redis/RedisMenuStockCacheService.java

```diff
@@ -0,0 +1,103 @@
+package camp.woowak.lab.infra.cache.redis;
+
+import org.redisson.api.RAtomicLong;
+import org.redisson.api.RLock;
+import org.redisson.api.RedissonClient;
+import org.springframework.stereotype.Service;
+
+import camp.woowak.lab.infra.cache.MenuStockCacheService;
+import camp.woowak.lab.infra.cache.exception.CacheMissException;
+import camp.woowak.lab.infra.cache.exception.RedisLockAcquisitionException;
+import camp.woowak.lab.menu.exception.NotEnoughStockException;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class RedisMenuStockCacheService implements MenuStockCacheService {
+	private final RedissonClient redissonClient;
+
+	public Long updateStock(Long menuId, Long stock) {
+		// redis 에 {메뉴ID: 재고} 가 있는 지 확인
+		try {
+			for (int retryCount = 0; retryCount < RedisCacheConstants.RETRY_COUNT; retryCount++) {
+				// 캐시에 재고가 있는 지 확인
+				RAtomicLong cachedStock = redissonClient.getAtomicLong(RedisCacheConstants.MENU_STOCK_PREFIX + menuId);
+				if (cachedStock.isExists()) { // 캐시에 재고가 있는 경우: 캐시 반환
+					return cachedStock.get();
+				}
+
+				// 캐시에 재고가 없는 경우: 락 확인
+				// 락이 걸려있지 않은 경우 락 걸고 {메뉴ID:재고} 캐시 생성
+				if (doWithLock(RedisCacheConstants.MENU_PREFIX + menuId,
+					() -> cachedStock.set(stock))) { // action 성공 시 재시도 탈출
+					break;
+				}
+			}
+
+			return stock;
+		} catch (Exception e) { // redis client 를 연결하지 못하면 넘어가도록 처리
+			log.error("[Redis Cache Error]: {}", e.getMessage());
+			return stock;
+		}
+	}
+
+	@Override
+	public Long addAtomicStock(Long menuId, int amount) {
+		RAtomicLong cachedStock = redissonClient.getAtomicLong(RedisCacheConstants.MENU_STOCK_PREFIX + menuId);
+		if (!cachedStock.isExists()) {    // 캐시 미스
+			throw new CacheMissException("메뉴 재고 캐시 미스");
+		}
+
+		// cache 원자적 재고감소
+		Long newStock = cachedStock.addAndGet(amount);
+
+		if (newStock < 0) {
+			// 원복
+			cachedStock.addAndGet(-amount);
+			throw new NotEnoughStockException("MenuId(" + menuId + ") 재고가 부족합니다.");
+		}
+		return newStock;
+	}
+
+	public boolean doWithLock(String key, Runnable runnable) {
+		RLock lock = redissonClient.getLock(RedisCacheConstants.LOCK_PREFIX + key);
+		boolean lockAcquired = false;
+		try {
+			lockAcquired = lock.tryLock(
+				RedisCacheConstants.LOCK_WAIT_TIME,
+				RedisCacheConstants.LOCK_LEASE_TIME,
+				RedisCacheConstants.LOCK_TIME_UNIT
+			);
+
+			if (!lockAcquired) {
+				return false;
+			}
+			runnable.run();
+			return true;
+		} catch (InterruptedException e) {
+			throw new RedisLockAcquisitionException("[redis lock] 락 획득에 실패했습니다.", e);
+		} finally {
+			if (lockAcquired) {
+				releaseLock(lock, RedisCacheConstants.LOCK_PREFIX + key);
+			}
+		}
+	}
+
+	private void releaseLock(RLock rLock, String key) {
+		try {
+			if (rLock.isHeldByCurrentThread()) {
+				rLock.unlock();
+				log.info("Released lock with key {}", key);
+			}
```

Comment:
```suggestion
			rLock.unlock();
			log.info("Released lock with key {}", key);
```

내부 코드를 뜯어보니 redis에 추가적인 네트워크 요청이 가는 것 같네요.
추가적인 검증이 필요한 상황이 아니면 제거하는 것도 좋을 것 같아요.

Reply:
감사합니다!